File: telegram-bot-core/src/main/java/ru/tggc/telegrambotcore/registry/HandlerScanner.java
package ru.tggc.telegrambotcore.registry;

import lombok.RequiredArgsConstructor;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.ListableBeanFactory;
import org.springframework.stereotype.Component;
import ru.tggc.telegrambotcore.annotation.handle.DefaultMessageHandle;

import java.lang.annotation.Annotation;
import java.lang.reflect.Method;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.regex.Pattern;

@Component
@Slf4j
@RequiredArgsConstructor
public class HandlerScanner {
    private final ListableBeanFactory beanFactory;

    @SneakyThrows
    public HandlerRegistryData scan(Class<? extends Annotation> annotationClass, Class<? extends Annotation> handlerClass) {
        if (annotationClass == null) {
            return null;
        }
        Map<String, RegisteredHandler> handlers = new ConcurrentHashMap<>();

        Method defaultMethod = null;
        Object defaultBean = null;

        Map<String, Object> handlerBeans = beanFactory.getBeansWithAnnotation(handlerClass);

        for (Object bean : handlerBeans.values()) {
            for (Method method : bean.getClass().getDeclaredMethods()) {
                if (method.isAnnotationPresent(annotationClass)) {
                    RegisteredHandler handler = new RegisteredHandler();
                    Annotation ann = method.getAnnotation(annotationClass);

                    String key = (String) ann.annotationType()
                            .getMethod("value")
                            .invoke(ann);

                    handler.setMethod(method);
                    handler.setBean(bean);

                    if (key.contains("${")) {
                        String regex = key.replaceAll("\\$\\{(\\w+)}", "(?<$1>\\\\S+)");
                        handler.setPattern(Pattern.compile(regex));
                    }

                    log.info("Registered handler '{}' -> {}.{}",
                            key,
                            bean.getClass().getSimpleName(),
                            method.getName()
                    );

                    handlers.put(key, handler);
                }

                if (method.isAnnotationPresent(DefaultMessageHandle.class)) {
                    if (defaultMethod != null) {
                        throw new IllegalStateException("Должен быть только один @DefaultMessageHandle");
                    }

                    defaultMethod = method;
                    defaultBean = bean;
                }
            }
        }

        return new HandlerRegistryData(
                handlers,
                defaultMethod,
                defaultBean
        );
    }
}

File: telegram-bot-core/src/main/kotlin/ru/tggc/telegrambotcore/registry/HandlerScanner.kt
package ru.tggc.telegrambotcore.registry

import io.github.oshai.kotlinlogging.KotlinLogging
import lombok.SneakyThrows
import org.springframework.beans.factory.ListableBeanFactory
import org.springframework.stereotype.Component
import java.util.concurrent.ConcurrentHashMap
import java.util.regex.Pattern

@Component
class HandlerScanner(private val beanFactory: ListableBeanFactory) {
    private val log = KotlinLogging.logger {}

    @SneakyThrows
    fun scan(annotationClass: Class<out Annotation>?, handlerClass: Class<out Annotation?>): HandlerRegistryData? {
        if (annotationClass == null) {
            return null
        }
        val handlers: MutableMap<String, RegisteredHandler> = ConcurrentHashMap()

        val handlerBeans: MutableMap<String, Any> = beanFactory.getBeansWithAnnotation(handlerClass)

        for (bean in handlerBeans.values) {
            for (method in bean.javaClass.declaredMethods) {
                if (method.isAnnotationPresent(annotationClass)) {
                    val handler = RegisteredHandler()
                    val ann: Annotation = method.getAnnotation(annotationClass)

                    val key = ann.annotationClass.java
                        .getMethod("value")
                        .invoke(ann) as String

                    handler.method = method
                    handler.bean = bean

                    if (key.contains($$"${")) {
                        val regex = key.replace("\\$\\{(\\w+)}".toRegex(), "(?<$1>\\\\S+)")
                        handler.pattern = Pattern.compile(regex)
                    }

                    log.info { "Registered handler '$key' -> ${bean.javaClass.simpleName}.${method.name}" }

                    handlers[key] = handler
                }
            }
        }

        return HandlerRegistryData(handlers)
    }
}
